designs/0004-closures-fun-types.md

```diff
@@ -0,0 +1,608 @@
+list followed by a function body.   For example, in `(real u) { return
+1 / (1 + exp(-u)); }`, the function argument list is `(real u)` and
+the function body is `{ return 1 / (1 + exp(-u)); }`.
+
+As a shorthand, a function argument list followed by an expression is
+taken to be shorthand for the function argument list followed by a
+function body returning that expression.  For example, the lambda
+expression
+
+```
+(real u) { return 1 / (1 + exp(-u)); }
+```
+
+may be replaced with the equivalent
+
+```
+(real u) 1 / 1 + exp(-u)
+```
+
+Lambda expressions behave like other expressions such as `2 + 2`, only
+they denote functions rather than values and have function types.
+
+
+## Changes to Standard Functions
+
+Allow functions to be defined in any scope.  Allow function
+definitions and lambda expressions to
+capture variables in scope by references.  
+
+Deprecate the functions block;  functions declared there can be
+declared in the transformed data block.  
+
+Standard function definitions of the form
+
+```
+T0 f(T1 x1, ..., TN xN) { ... }
+```
+
+may be replaced with
+
+```
+T0(T1, ..., TN) f = (T1 x1, ..., TN xN) { ... }
+```
+
+## Variable declarations
+
+Variable declarations for functions follows the usual syntax, with a
+sized type required for block variables, an unsized type for function
+arguments, and either a sized (as of 2.20) or unsized (future) type
+for a local variable declaration.
+
+
+## Assignment
+
+Nothing changes conceptually about assignment.  The right-hand side
+type must still be assignable to the left-hand side type, which for
+functions, means they have the same function type.  The only question
+is whether to require strict matching of types or support full
+covariance and contravariance.  
+
+## Implementation
+
+Lambdas can be mapped directly to C++ lambdas with default reference
+closures.  This will work because the scope of the compiled C++ is the
+same as that of the Stan program.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Like all new features, more coding, testing, doc, and ongoing maintenance.
+
+Higher-order programming is hard.  If we use too much of it, we risk
+alienating users who are looking for something simpler.
+
+There is no I/O for types with functions in them, as there is no way
+to save a function.  This will complicate various test programs and
+I/O programs involving typed variables.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This is really two proposals, one for closures (for capturing
+variables in scope) and one for lambdas (for anonymous functions).
+Although they naturally go together, it would be possible to adopt
+closures without lambdas or vice-versa.  
+
+Disallowing lambdas would complicate standard functional idioms like
+maps, which rely on simple inline anonymous lambdas.
+
+Disallowing closures requires passing all values to higher order
+functions as arrays along with the packing and unpacking required.
```

Comment:
Does this proposal immediately allow us to simplify the signatures of the ODE-solver and map_rect significantly, if we ask people to pass in a closure rather than a mere function?

Reply:
I think so, if we implement closures via something similar to lambda lifting but with automatic packing and unpacking of data and parameters.